pkg/agent/api/grpc/hyperstart.proto

```diff
@@ -0,0 +1,135 @@
+syntax = "proto3";
+
+package grpc;
+
+import "oci.proto";
+import "google/protobuf/empty.proto";
+
+// To generate hyperstart.pb.go, run:
+// $ go get -u github.com/google/protobuf
+// $ protoc -I=$GOPATH/src/github.com/google/protobuf/src/ -I. --go_out=plugins=grpc:. hyperstart.proto oci.proto
+
+// unstable
+service HyperstartService {
+	// execution
+	rpc CreateContainer(CreateContainerRequest) returns (google.protobuf.Empty);
+	rpc StartContainer(StartContainerRequest) returns (google.protobuf.Empty);
+	rpc ExecProcess(ExecProcessRequest) returns (google.protobuf.Empty);
+	rpc SignalProcess(SignalProcessRequest) returns (google.protobuf.Empty);
+	rpc WaitProcess(WaitProcessRequest) returns (WaitProcessResponse); // wait & reap like waitpid(2)
+
+	// stdio
+	rpc WriteStdin(WriteStreamRequest) returns (WriteStreamResponse);
+	rpc ReadStdout(ReadStreamRequest) returns (ReadStreamResponse);
+	rpc ReadStderr(ReadStreamRequest) returns (ReadStreamResponse);
+	rpc CloseStdin(CloseStdinRequest) returns (google.protobuf.Empty);
+	rpc TtyWinResize(TtyWinResizeRequest) returns (google.protobuf.Empty);
+
+	// misc (TODO: some rpcs can be replaced by hyperstart-exec)
+	rpc StartSandbox(StartSandboxRequest) returns (google.protobuf.Empty);
+	rpc DestroySandbox(DestroySandboxRequest) returns (google.protobuf.Empty);
+	rpc UpdateInterface(UpdateInterfaceRequest) returns (google.protobuf.Empty);
+	rpc AddRoute(AddRouteRequest) returns (google.protobuf.Empty);
+	rpc OnlineCPUMem(OnlineCPUMemRequest) returns (google.protobuf.Empty);
+}
+
```

Comment:
```proto
message Storage {
    string driver = 1;   // empty in most cases. it will support "drbd", "bcache" ...
    string source = 2; // "/dev/sdb", "/dev/disk/by-scsi/xxxx", "none",...
    string fstype = 3; //"xfs", "ext4" etc. for block dev, or "9p" for share, or  "tmpfs" for shared /dev/shm for all containers ...
    repeat string options = 4;
    string mount_point = 5; // mount_point is only visible by VM, not by containers. This mount point can be used on oci.Mount.Source as "/Stroage/mount/point/{rootfs|_data}".
}
```
I think this Storage + oci.Mount can reassemble current hyperstart's storage model.

Reply:
@laijs Would you mind describing use cases for this `Storage` payload?

Reply:
@sameo In short, for the backward compatibility to current hyperd/hyperstart.

The `Storage` derives from the volumes of hyperd and runv, in which we could insert a block device (the `source`, loop device or ceph rbd is ok as well) as a volume of container (as filesystem `fstype`, to the `mount_point`).

According to assumptions of the WIP CSI, the block devices would be inserted as block devices instead of mounted filesystems, i.e., for most current container images, there need a k8s init container to mount the block devices before the devices could be accessed.

The `Storage` struct here could be think as a shortcut for the volume init container behavior because the block devices are much more common in a virtualized world. Then we don't need run a volume init container before launch a container, especially for the case not working with k8s.

Any additional explanation, @laijs ?

Reply:
@gnawux explained it in high level view. Here I focus on low level view:

We need the initial storage config for mounting the 9p which can be configured by this Storage in StartSandboxRequest.

oci.Spec can't ask you "mount /dev/sda1 mount to /kata/storage/xxxx and use /kata/storage/xxxx/rootfs as container rootfs".

oci.Spec can't ask you "mount /dev/sdd mount to /kata/storage/yyy and use /kata/storage/xxxx/_data as one of the container volume".

oci.Mount can't ask you mount a tmpfs on /kata/storage/shm, and bind it to all container's /dev/shm.

Reply:
@laijs I was about to ask for a `shareDir` parameter for `StartSandbox` since we need a way to pass some file and rootfs through 9pfs. Your `Storage` structure looks good because it is a generic way to pass some things to the VM through any type of filesystem that need to be shared accross all containers, and that are not hotpluggable actually.

Reply:
Also, to give some more details here, we cannot completely rely on block devices to pass rootfs to the VM. In case of `overlay`, we don't have a block device and I don't think we want to spend a large amount of time preparing a block device based on this, it would consume too much time. Also I am not sure how it would work to reflect the changes on the block device on the overlay layers.
That's why having a mount point that can be shared when starting the VM would be the best option to support multiple cases.

Reply:
@laijs we have renamed the `driver` field to `format` in runV, and the content should be stuff like `raw`, `qcow2`, `dir`... we don't mind it is `bcache` device or a loop file if they could be treat as a block device.

Reply:
We use block devices for passing rootfs to the VM in several cases. Host storage drivers such as devmapper, (cow-)rawblock, ceph(hyper.sh) fit well into runv/hyperstart. In these cases, block devices containing container rootfs or volume are hotplug into the VM. In near future, DAX is going to be supported which also relies on block devices.

Only when the host storage drivers are overlayfs/aufs/btrfs, 9pfs is used for passing rootfs.
So both ways(block&sharedfs) are supported. @sboeuf